src/Roslyn.Diagnostics.Analyzers/Core/AbstractCreateTestAccessor`1.cs

```diff
@@ -1,70 +1,82 @@
+
         protected abstract SyntaxNode GetTypeDeclarationForNode(SyntaxNode reportedNode);
 
-        public override async Task RegisterCodeFixesAsync(CodeFixContext context)
+        public override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
         {
+            var type = await GetRelevantTypeFromHeaderAsync(context).ConfigureAwait(false);
+            if (type is null)
+                return;
+
             var semanticModel = await context.Document.GetSemanticModelAsync(context.CancellationToken).ConfigureAwait(false);
+            var typeSymbol = (INamedTypeSymbol)semanticModel.GetDeclaredSymbol(type, context.CancellationToken);
+            if (!IsClassOrStruct(typeSymbol))
+                return;
+
+            if (typeSymbol.GetTypeMembers(TestAccessorHelper.TestAccessorTypeName).Any())
+                return;
+
+            var location = typeSymbol.Locations.FirstOrDefault(location => location.IsInSource && Equals(location.SourceTree, semanticModel.SyntaxTree));
+            if (location is null)
+                return;
 
-            foreach (var diagnostic in context.Diagnostics)
-            {
-                var syntaxTree = diagnostic.Location.SourceTree;
-                var syntaxRoot = await syntaxTree.GetRootAsync(context.CancellationToken).ConfigureAwait(false);
-                var reportedNode = syntaxRoot.FindNode(diagnostic.Location.SourceSpan, getInnermostNodeForTie: true);
-                var typeDeclaration = GetTypeDeclarationForNode(reportedNode);
-                var type = (ITypeSymbol)semanticModel.GetDeclaredSymbol(typeDeclaration, context.CancellationToken);
-                if (type.GetTypeMembers(CreateTestAccessor.TestAccessorTypeName).Any())
-                {
-                    continue;
-                }
-
-                context.RegisterCodeFix(
-                    CodeAction.Create(
-                        RoslynDiagnosticsAnalyzersResources.CreateTestAccessorMessage,
-                        cancellationToken => CreateTestAccessorAsync(context.Document, diagnostic, cancellationToken),
-                        nameof(CreateTestAccessorFixer)),
-                    diagnostic);
-            }
+            context.RegisterRefactoring(
+                CodeAction.Create(
+                    RoslynDiagnosticsAnalyzersResources.CreateTestAccessorMessage,
+                    cancellationToken => CreateTestAccessorAsync(context.Document, location.SourceSpan, cancellationToken),
+                    nameof(AbstractCreateTestAccessor<TTypeDeclarationSyntax>)));
         }
 
-        private async Task<Document> CreateTestAccessorAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        private async Task<TTypeDeclarationSyntax?> GetRelevantTypeFromHeaderAsync(CodeRefactoringContext context)
+        {
+            var type = await context.TryGetRelevantNodeAsync<TTypeDeclarationSyntax>(RefactoringHelpers).ConfigureAwait(false);
```

Comment:
Can we inline this call and remove `GetRelevantTypeFromHeaderAsync`?

Reply:
Done